lms/djangoapps/program_enrollments/api/v1/views.py

```diff
@@ -0,0 +1,249 @@
+        """
+        self.check_existence_and_membership()
+        results = {}
+        seen_student_keys = set()
+        enrollments = []
+
+        if not isinstance(request.data, list):
+            raise ValidationError("invalid enrollment record")
+        if len(request.data) > 25:
+            return Response(
+                'enrollment limit 25', HTTP_413_REQUEST_ENTITY_TOO_LARGE
+            )
+
+        try:
+            for enrollment_request in request.data:
+                error_status = self.check_enrollment_request(enrollment_request, seen_student_keys)
+                if error_status:
+                    results[enrollment_request["student_key"]] = error_status
+                else:
+                    enrollments.append(enrollment_request)
+        except (KeyError, ValidationError, TypeError):
+            return Response('invalid enrollment record', HTTP_422_UNPROCESSABLE_ENTITY)
+
+        program_enrollments = self.get_existing_program_enrollments(enrollments)
+        for enrollment in enrollments:
+            student_key = enrollment["student_key"]
+            if student_key in results and results[student_key] == CourseEnrollmentResponseStatuses.DUPLICATED:
+                continue
+            results[student_key] = self.enroll_learner_in_course(enrollment, program_enrollments)
+
+        good_count = sum([1 for _, v in results.items() if v not in CourseEnrollmentResponseStatuses.ERROR_STATUSES])
+        if not good_count:
+            return Response(results, HTTP_422_UNPROCESSABLE_ENTITY)
+        if good_count != len(results):
+            return Response(results, HTTP_207_MULTI_STATUS)
+        else:
+            return Response(results)
+
+    def check_enrollment_request(self, enrollment, seen_student_keys):
+        """
+        Checks that the given enrollment record is valid and hasn't been duplicated
+        """
+        student_key = enrollment['student_key']
+        if student_key in seen_student_keys:
+            return CourseEnrollmentResponseStatuses.DUPLICATED
+        seen_student_keys.add(student_key)
+        enrollment_serializer = ProgramCourseEnrollmentRequestSerializer(data=enrollment)
+        try:
+            enrollment_serializer.is_valid(raise_exception=True)
+        except ValidationError as e:
+            if enrollment_serializer.has_invalid_status():
+                return CourseEnrollmentResponseStatuses.INVALID_STATUS
+            else:
+                raise e
+
+    def get_existing_program_enrollments(self, enrollments):
+        """
+        Parameters:
+            - enrollments: A list of enrollment requests
+        Returns:
+            - Dictionary mapping all student keys in the enrollment requests
+              to that user's existing program enrollment <self.program>
+        """
+        external_user_keys = [e["student_key"] for e in enrollments]
+        existing_enrollments = ProgramEnrollment.objects.filter(
```

Comment:
I'm not sure if we support this but we might want to consider what if a user is enrolled in multiple programs here

Reply:
I don't know if that's something we need to worry about but i'll add a filter for program uuid just to be doubly sure